### ISSUE TITLE
Error when creating branch protection

### ISSUE THREAD

**Issue by luisdavim** [google/go-github#2467]
Since we updated our code to use the latest version of this library, we can't create a branch protection that requires branches to be up to date before merging but don't require any status checks:

```go
func (r *RepoSettings) protectionRequestOptions() *github.ProtectionRequest {
	return &github.ProtectionRequest{
		EnforceAdmins: r.EnforceAdmins,
		RequiredPullRequestReviews: &github.PullRequestReviewsEnforcementRequest{
			DismissStaleReviews:          r.DismissStaleReviews,
			RequiredApprovingReviewCount: r.RequiredApprovingReviewCount,
			RequireCodeOwnerReviews:      r.RequireCodeOwnerReviews,
		},
		RequiredStatusChecks: &github.RequiredStatusChecks{
			Strict: true,
		},
	}
}
```

The above results in an error:

```
PUT https://api.github.com/repos/****/****/branches/main/protection: 422 Invalid request.

No subschema in "anyOf" matched.
No subschema in "oneOf" matched.
Not all subschemas of "allOf" matched.
For 'anyOf/1', {"strict"=>true} is not a null. []
```

Such rule is possible to be created from the UI:
<img width="772" alt="image" src="https://user-images.githubusercontent.com/173006/190856798-25e91666-95dc-491a-a71c-f7f454db4786.png">

This is also affecting the Github Terraform provider: https://github.com/integrations/terraform-provider-github/issues/1147

**Comment by luisdavim** [comment]
Looking at https://github.com/orgs/community/discussions/24642 removing `omitempty` from the `Checks` property could solve the problem but I guess that for that you'll have to fully deprecate the `Contexts` or maybe you could force `Checks` to be null if `len(Contexts) > 0 && len(Contexts) == 0`

**Comment by luisdavim** [comment]
I've tested this locally, removing `omitempty` from the `Checks` property in https://github.com/google/go-github/blob/master/github/repos.go#L908 and passing:

```go
github.ProtectionRequest{
	EnforceAdmins: r.EnforceAdmins,
	RequiredPullRequestReviews: &github.PullRequestReviewsEnforcementRequest{
		DismissStaleReviews:          r.DismissStaleReviews,
		RequiredApprovingReviewCount: r.RequiredApprovingReviewCount,
		RequireCodeOwnerReviews:      r.RequireCodeOwnerReviews,
	},
	RequiredStatusChecks: &github.RequiredStatusChecks{
		Strict: true,
		Checks: []*github.RequiredStatusCheck{},
	},
}
```

Works, and you might not need to fully deprecate `Contexts` as if `Checks` is not set it's passed as `null` and that should be ok if `Contexts` is set...

**Comment by luisdavim** [comment]
Thank you very much for the quick turnaround on this @gmlewis 🚀 
Any chances of getting a new tag released so we can get the fix downstream?

**Comment by gmlewis** [comment]
> Thank you very much for the quick turnaround on this @gmlewis rocket Any chances of getting a new tag released so we can get the fix downstream?

Yes, I can work on a new release today.

**Comment by luisdavim** [comment]
Awesome, thanks. Really appreciate it 👍

**Comment by gmlewis** [comment]
@luisdavim - this issue has now been covered in the latest release:
https://github.com/google/go-github/releases/tag/v47.1.0

**Comment by alock** [comment]
When attempting to update I ran into the same issue as https://github.com/integrations/terraform-provider-github/issues/1307.  It appears that golang is setting the value to `nil` and not `null` as the github api requires.  Is there a simple way to override that before the value is sent?

```For 'anyOf/1', {"strict"=>true, "contexts"=>["build"], "checks"=>nil} is not a null. []```

**Comment by gmlewis** [comment]
@luisdavim - do you have any insights to this new issue?

@alock - just for the record, this repo is now sending an empty `[]` which that error message is reporting as `nil`, as evidenced by our unit test here:
https://github.com/google/go-github/commit/23a2636cfa918c26a5ce2e7656ebeb982334a850#diff-1da0ffca0653d8b270e435e5e5ef6615b547c752c60b45df54c7e4307ece89e9R1589

**Comment by luisdavim** [comment]
I think the terraform provider still needs to be updated to use a version of this lib that contains the fix....

**Comment by gmlewis** [comment]
> I think the terraform provider still needs to be updated to use a version of this lib that contains the fix....

Ah, thank you, @luisdavim !

Then I'll close this issue to hopefully reduce the confusion... but please report back if you feel we introduced a regression.

**Comment by TheQueenIsDead** [comment]
> I think the terraform provider still needs to be updated to use a version of this lib that contains the fix....

It looks like the Terraform provider [updated to go-github v48.1.0 17 days ago](https://github.com/integrations/terraform-provider-github/commit/bb0f993d8510d1673ac5bc47dd8b3455d5f28312#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R8), and published those changes in their [v5.11.0 release](https://github.com/integrations/terraform-provider-github/releases/tag/v5.11.0)

However I still receive the same issue attempting to apply:
```terraform
╷
│ Error: PUT https://api.github.com/repos/<org>/status/branches/<b>/protection: 422 Invalid request.
│ 
│ No subschema in "anyOf" matched.
│ For 'properties/checks', nil is not an array.
│ Not all subschemas of "allOf" matched.
│ For 'anyOf/1', {"strict"=>false, "checks"=>nil} is not a null. []
│ 
│   with module.repository["<service>"].github_branch_protection_v3.branch_protection[1],
│   on .terraform/modules/repository/main.tf line 240, in resource "github_branch_protection_v3" "branch_protection":
│  240: resource "github_branch_protection_v3" "branch_protection" {
│ 
╵
╷
│ Error: PUT https://api.github.com/repos/<org>/status/branches/<a>/protection: 422 Invalid request.
│ 
│ No subschema in "anyOf" matched.
│ For 'properties/checks', nil is not an array.
│ Not all subschemas of "allOf" matched.
│ For 'anyOf/1', {"strict"=>false, "contexts"=>["go / Build"], "checks"=>nil} is not a null. []
│ 
│   with module.repository["<service>"].github_branch_protection_v3.branch_protection[0],
│   on .terraform/modules/repository/main.tf line 240, in resource "github_branch_protection_v3" "branch_protection":
│  240: resource "github_branch_protection_v3" "branch_protection" {
│ 
╵

```

Versions as such, is there something I'm missing here?
```
$ terraform version
Terraform v1.3.5
on linux_amd64
+ provider registry.terraform.io/integrations/github v5.11.0
```

Any help appreciated 🙏

**Comment by gmlewis** [comment]
Reopening but need assistance on this one from Terraform users.

**Comment by luisdavim** [comment]
The fix will have to be on the terraform provider side, to pass an empty array instead of nothing, the alternative could be to add logic to this library to set an empty array when it's `nil`....

**Comment by TheQueenIsDead** [comment]
> I've tested this locally, removing `omitempty` from the `Checks` property in https://github.com/google/go-github/blob/master/github/repos.go#L908 and passing:
> 
> ```go
> github.ProtectionRequest{
> 	EnforceAdmins: r.EnforceAdmins,
> 	RequiredPullRequestReviews: &github.PullRequestReviewsEnforcementRequest{
> 		DismissStaleReviews:          r.DismissStaleReviews,
> 		RequiredApprovingReviewCount: r.RequiredApprovingReviewCount,
> 		RequireCodeOwnerReviews:      r.RequireCodeOwnerReviews,
> 	},
> 	RequiredStatusChecks: &github.RequiredStatusChecks{
> 		Strict: true,
> 		Checks: []*github.RequiredStatusCheck{},
> 	},
> }
> ```
> 
> Works, and you might not need to fully deprecate `Contexts` as if `Checks` is not set it's passed as `null` and that should be ok if `Contexts` is set...
> 
> I've opened #2468 to remove the `omitempty`.

Oh I see, per your comment above. Apologies for missing that. I'll open a PR downstream to try adding an empty array to the request in their utility file :+1: 
https://github.com/integrations/terraform-provider-github/blob/main/github/resource_github_branch_protection_v3_utils.go#L178

The idea to handle the nil gracefully upstream is a nice one, though you'd know best if there might be any interesting ramifications of that. Thanks for the response!

**Comment by luisdavim** [comment]
Do keep on mind that if `contexts` is set then `checks` should be null. Using `contexts` is deprecated and terraform shouldn't be using it but I don't know if it is.

**Comment by gmlewis** [comment]
OK, then I think we are good to go and I'll close this issue (again).
Hopefully we can leave it closed this time.

**Comment by emmahsax** [comment]
I'm trying to run v57.0.0 of this, and I'm still getting this issue:
```
No subschema in "anyOf" matched.
No subschema in "oneOf" matched.
Not all subschemas of "allOf" matched.
For 'anyOf/1', {"strict"=>true} is not a null. []
```

This is what our branch protection looks like:
```
defaultBranchProtection := &github.ProtectionRequest{
    RequiredStatusChecks: &github.RequiredStatusChecks{
	Checks: []*github.RequiredStatusCheck{},
	Strict: true,

    },
}
```

**Comment by gmlewis** [comment]
Reopening.

**Comment by emmahsax** [comment]
I did find that [v57.0.0](https://pkg.go.dev/github.com/google/go-github/v57/github#RequiredStatusChecks) has the `omitempty` on `Checks`, while [v56.0.0](https://pkg.go.dev/github.com/google/go-github/v56/github#RequiredStatusChecks) does not.

**Comment by KatharinaSick** [comment]
fyi: I'm getting the same issue. I tried to work around it but couldn't find a way.

**Comment by gmlewis** [comment]
This issue is apparently in direct conflict with #2976.

We would appreciate it if one of our amazing contributors would please volunteer to take a deep dive here and see why #2467 and #2976 are at opposition with one another and see if they can come up with a fix that solves both issues.

**Comment by KatharinaSick** [comment]
Thank you! :) 
Is there anything I can do to work around this in the meantime?

**Comment by gmlewis** [comment]
> Thank you! :) Is there anything I can do to work around this in the meantime?

I would recommend using the version that works for you for your particular use case, since we appear to have different use cases apparently.

This difference in use cases will be part of what needs to be determined with a deep dive into fixing this issue.

**Comment by gmlewis** [comment]
Please check out this comment: https://github.com/google/go-github/issues/2976#issuecomment-1883775544

**Comment by emmahsax** [comment]
> I think, if the `contexts` field is deprecated, then it should also be deprecated here. I _think_, if using contexts still, this is what should be done:
>
> ```go
> &github.RequiredStatusChecks{
>   Checks: []*github.RequiredStatusCheck{},
>   Contexts: []string{"hello-world"},
>   Strict: true,
> },
> ```
> 
> Basically, `Checks` should be required, and if the user would like to use `Contexts` instead, they must make `Checks` an empty slice. I've tested this with v57 of the go API and it works well and adds the `hello-world` context properly.
> 
> If people agree with this consensus, I'd be fine making a PR.

_Originally posted by @emmahsax in https://github.com/google/go-github/issues/2976#issuecomment-1934769027_

**Comment by emmahsax** [comment]
@KatharinaSick I've come with two potential workarounds that could work for your situation. If neither of them work, maybe I can try to help you find a solution. The workarounds are written here: https://github.com/google/go-github/issues/2976#issuecomment-1935057358

**Comment by gmlewis** [comment]
Fixed by: #3070.
